distsql,jobs: propagate 'job' log tag and pprof label to remote nodes

lidorcarmel opened the pull request:
Previously, the job log tags and the recently added job pprof labels were only used locally and not sent to remote nodes. Which means that, for example, pprof did not show the job ids of jobs that consume cpu but were started on another node.

This PR propagates the job tag to the remote nodes, and there it adds that tag to be used when logging. It also adds the same tag as a pprof label on the remote node. With this change we no longer need to set the tags that were added in https://github.com/cockroachdb/cockroach/pull/77397, therefore those are reverted here.

Note that this PR changes the job log tag from `job=<job id>` to `job=<job type> id=<job id>`, to be the same as the pprof label.

See this gist for a before/after example https://gist.github.com/lidorcarmel/2c147b76d814e29682cccbdcadb5d7bb.

And here is pprof graph example with those tags:

<img width="2531" alt="pprof_with_job_tags" src="https://user-images.githubusercontent.com/51982110/201506953-acf2675d-28a2-471e-9ebf-5441b4728978.png">

There is another commit in this PR which is fixing a small bug in the recently added pprofutil package.

Epic: None

Release note: None

Commit:
distsql,jobs: propagate 'job' log tag and pprof label to remote nodes

Previously, the recently added job log tag and pprof label
were only used locally and not sent to remote nodes. Which means that,
for example, pprof did not show the job ids of jobs that consume cpu
but were started on another node.

This PR propagates the job tag to the remote nodes, and there it
adds that tag to be used when logging. It also adds the same tag as
a pprof label on the remote node. With this change we no longer need
to set the tags that were added in #77397, but we cannot remove those
yet because we want the old job tags to still be sent and applied
during upgrades. Once 23.1 is out, we can remove the old implementation
from 23.2.

Note that this PR changes the job log tag from `job=<job id>` to
`job=<job type> id=<job id>`, to be the same as the pprof label.

Epic: None

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/91630)
<!-- Reviewable:end -->

Commit:
pprofutil: return valid ctx and func when no log tags are found

If we don't have log tags in the context we return invalid context and undo
func and then the caller might crash.

Return an empty func and the same context.

Add a couple of basic tests.

Epic: None

Release note: None

lidorcarmel commented:
cc @tbg FYI.

yuzefovich reviewed (commented):
Should we split up the first commit into a separate PR to be backported to 22.2? Or do we plan to backport the whole change there?

> Reviewed 4 of 4 files at r1, 22 of 23 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91630#-:-NGxDDWD96Xx8d5L8B5S:b-1eg73l)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @HonoreDB, and @lidorcarmel)

___
*[`-- commits` line 18 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91630#-NGxDMfs25AsaBDz-cGJ):*
nit: `s/tagand/tag and/`.
___
*[`pkg/sql/distsql_running.go` line 410 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91630#-NGxEkpqCt_rkLweiCEa:-NGxEkpqCt_rkLweiCEb:baw8cy5) ([raw file](https://github.com/cockroachdb/cockroach/blob/92bf199109de12ca532b15d57063f41de5180e8c/pkg/sql/distsql_running.go#L410)):*
> ```Go
> 		return ""
> 	}
> 	setupReq := execinfrapb.SetupFlowRequest{
> ```

nit: could you please add the TODO for me here:
```
	// TODO(yuzefovich): avoid populating some fields of the SetupFlowRequest
	// for local plans.
```
___
*[`pkg/sql/execinfra/server_config.go` line 304 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91630#-NGxFRMt4E3x68kMnC3o:-NGxFRMt4E3x68kMnC3p:b-14yfr7) ([raw file](https://github.com/cockroachdb/cockroach/blob/92bf199109de12ca532b15d57063f41de5180e8c/pkg/sql/execinfra/server_config.go#L304)):*
> ```Go
> 	// VerifyContext, when non-nil, is called by the execinfrapb.DistSQLServer
> 	// when responding to SetupFlow RPCs.
> 	VerifyContext func(ctx context.Context)
> ```

Is it possible to expand `SetupFlowCb` to do what `VerifyContext` does (or in the opposite direction)? I think the current single usage of `SetupFlowCb` would be ok if it was called right before `RunFlow`.


<!-- Sent from Reviewable.io -->

lidorcarmel reviewed (commented):
thanks for asking! we just discussed it this morning, we probably want to backport everything to 22.2.x. And the pprofutil code touched by the first commit is all new code, so no need to backport to earlier releases.

even if we don't backport, I guess I'm not too worried about backporting the small pprofutil commit because the other 2 usages of `SetProfilerLabelsFromCtxTags` always add a tag first and should succeed.

but with the current pr, in mixed version the new<->old servers will not propagate log tags, so I'll change this PR to still send the job ID but add it as a tag only if it's not there already. cc @dt 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91630#-:-NH6KMYo2o1mFgiX8pm-:b-5ytw7i)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @HonoreDB, and @yuzefovich)

___
*[`-- commits` line 18 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91630#-NGxDMfs25AsaBDz-cGJ):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: `s/tagand/tag and/`.
</blockquote></details>

done
___
*[`pkg/sql/distsql_running.go` line 410 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91630#-NGxEkpqCt_rkLweiCEa:-NH2kVlg7KyD9_7M5K9e:b-wwu3pa) ([raw file](https://github.com/cockroachdb/cockroach/blob/92bf199109de12ca532b15d57063f41de5180e8c/pkg/sql/distsql_running.go#L410)):*
> // TODO(yuzefovich): avoid populating some fields of the SetupFlowRequest
> 	// for local plans.

of course..
___
*[`pkg/sql/execinfra/server_config.go` line 304 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91630#-NGxFRMt4E3x68kMnC3o:-NH2uhVZE4MdmtjP06Gr:b-hsklmq) ([raw file](https://github.com/cockroachdb/cockroach/blob/92bf199109de12ca532b15d57063f41de5180e8c/pkg/sql/execinfra/server_config.go#L304)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Is it possible to expand `SetupFlowCb` to do what `VerifyContext` does (or in the opposite direction)? I think the current single usage of `SetupFlowCb` would be ok if it was called right before `RunFlow`.
</blockquote></details>

unfortunately if we call that callback after setting up the flow the test will fail, because we see an inflight flow here:
https://github.com/cockroachdb/cockroach/blob/0c1095e31cf93ea7f177f8bf1750ebab188e02d7/pkg/sql/distsql_running_test.go#L667

if there's an easy way around it I can try.
otherwise I can s/VerifyContext/PostSetupFlowCb/ 


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91630)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @HonoreDB, and @lidorcarmel)

___
*[`pkg/sql/execinfra/server_config.go` line 304 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91630#-NGxFRMt4E3x68kMnC3o:-NH6esWP9pp7qdDfv2-b:bb00lj5) ([raw file](https://github.com/cockroachdb/cockroach/blob/92bf199109de12ca532b15d57063f41de5180e8c/pkg/sql/execinfra/server_config.go#L304)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

unfortunately if we call that callback after setting up the flow the test will fail, because we see an inflight flow here:
https://github.com/cockroachdb/cockroach/blob/0c1095e31cf93ea7f177f8bf1750ebab188e02d7/pkg/sql/distsql_running_test.go#L667

if there's an easy way around it I can try.
otherwise I can s/VerifyContext/PostSetupFlowCb/ 
</blockquote></details>

Hm, I'd like to not introduce another testing knob if we can. I just pushed another commit to your branch that seems to work (I stressed it for a couple of minutes locally). Let me know what you think of that change.


<!-- Sent from Reviewable.io -->

lidorcarmel reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91630)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @HonoreDB, and @yuzefovich)

___
*[`pkg/sql/execinfra/server_config.go` line 304 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91630#-NGxFRMt4E3x68kMnC3o:-NH77d4SC1VPK8TdS6bD:b-mq9b2y) ([raw file](https://github.com/cockroachdb/cockroach/blob/92bf199109de12ca532b15d57063f41de5180e8c/pkg/sql/execinfra/server_config.go#L304)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Hm, I'd like to not introduce another testing knob if we can. I just pushed another commit to your branch that seems to work (I stressed it for a couple of minutes locally). Let me know what you think of that change.
</blockquote></details>

ah, I see why what I did didn't work, I didn't do the cleanup. your commit lgtm, thanks!


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> it might good to get a sign off from someone else too.

> Reviewed 4 of 5 files at r4, 16 of 16 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91630#-:-NHCAGclDGM7zYLh8gLZ:b5tavr)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru, @dt, and @HonoreDB)



<!-- Sent from Reviewable.io -->

dt commented on pkg/sql/execinfrapb/processors_bulk_io.proto:
```diff
@@ -37,8 +37,9 @@ import "cloud/cloudpb/external_storage.proto";
 // descriptor in the database, and doesn't emit any rows nor support
 // any post-processing.
 message BackfillerSpec {
+  // TODO(lidor): this is not needed in 23.2, the new way to send the job tag is
```
minor nit: rather than "in 23.2", I'd say "when interoperability with 22.2  is dropped" or something, which may or may not be 23.2 but we'd be trying to predict the future if we specify which version will drop backwards compat with 22.2.

lidorcarmel replied:
Done, thanks!

lidorcarmel commented:
@yuzefovich PTAL.
`TestSetupFlowAfterDrain` failed because we double `Cleanup()`, so I reverted the small change in `RunFlow()`  so that it will do the cleanup as it did before (only if the error is `stop.ErrUnavailable`). And I added a `Cleanup` when the `SetupFlowCb` is set.

yuzefovich reviewed (approved):
I see, makes sense, LGTM

> Reviewed 16 of 16 files at r6, 11 of 11 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91630#-:-NHRQBPo4AgYZmsgDGYo:b-fcusj4)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru and @HonoreDB)



<!-- Sent from Reviewable.io -->

lidorcarmel commented:
Thanks!
bors r+
